# **Análisis Exploratorio De Datos**

In [1]:
import pandas as pd

In [2]:
# Leer primero archivo
archivo1 = r"C:\Users\Luis Sazo\Desktop\Análisis Notas Estudiantes\Data\Original_data_with_more_rows.csv"
DF_Uno = pd.read_csv(archivo1)
DF_Uno

Unnamed: 0  Gender EthnicGroup          ParentEduc     LunchType  \
0               0  female     group B   bachelor's degree      standard   
1               1  female     group C        some college      standard   
2               2  female     group B     master's degree      standard   
3               3    male     group A  associate's degree  free/reduced   
4               4    male     group C        some college      standard   
...           ...     ...         ...                 ...           ...   
30636         995    male     group C    some high school      standard   
30637         996    male     group E  associate's degree  free/reduced   
30638         997    male     group C        some college      standard   
30639         998    male     group A    some high school  free/reduced   
30640         999  female     group D  associate's degree      standard   

        TestPrep  MathScore  ReadingScore  WritingScore  
0           none         72            72            74  
1      completed         69            90            88  
2           none         90            95            93  
3           none         47            57            44  
4           none         76            78            75  
...          ...        ...           ...           ...  
30636       none         56            47            51  
30637       none         74            75            72  
30638       none         36            29            27  
30639  completed         43            34            39  
30640       none         52            68            66  

[30641 rows x 9 columns]

In [44]:
DF_Uno["EthnicGroup"].unique()


array(['group B', 'group C', 'group A', 'group D', 'group E'],
      dtype=object)

In [3]:
# Leer segundo archivo
archivo2 = r"C:\Users\Luis Sazo\Desktop\Análisis Notas Estudiantes\Data\Expanded_data_with_more_features.csv"
DF_Dos = pd.read_csv(archivo2)
DF_Dos

Unnamed: 0  Gender EthnicGroup          ParentEduc     LunchType  \
0               0  female         NaN   bachelor's degree      standard   
1               1  female     group C        some college      standard   
2               2  female     group B     master's degree      standard   
3               3    male     group A  associate's degree  free/reduced   
4               4    male     group C        some college      standard   
...           ...     ...         ...                 ...           ...   
30636         816  female     group D         high school      standard   
30637         890    male     group E         high school      standard   
30638         911  female         NaN         high school  free/reduced   
30639         934  female     group D  associate's degree      standard   
30640         960    male     group B        some college      standard   

        TestPrep ParentMaritalStatus PracticeSport IsFirstChild  NrSiblings  \
0           none             married     regularly          yes         3.0   
1            NaN             married     sometimes          yes         0.0   
2           none              single     sometimes          yes         4.0   
3           none             married         never           no         1.0   
4           none             married     sometimes          yes         0.0   
...          ...                 ...           ...          ...         ...   
30636       none              single     sometimes           no         2.0   
30637       none              single     regularly           no         1.0   
30638  completed             married     sometimes           no         1.0   
30639  completed             married     regularly           no         3.0   
30640       none             married         never           no         1.0   

      TransportMeans WklyStudyHours  MathScore  ReadingScore  WritingScore  
0         school_bus            < 5         71            71            74  
1                NaN         5 - 10         69            90            88  
2         school_bus            < 5         87            93            91  
3                NaN         5 - 10         45            56            42  
4         school_bus         5 - 10         76            78            75  
...              ...            ...        ...           ...           ...  
30636     school_bus         5 - 10         59            61            65  
30637        private         5 - 10         58            53            51  
30638        private         5 - 10         61            70            67  
30639     school_bus         5 - 10         82            90            93  
30640     school_bus         5 - 10         64            60            58  

[30641 rows x 15 columns]

In [45]:
DF_Dos["EthnicGroup"].unique()

array([nan, 'group C', 'group B', 'group A', 'group D', 'group E'],
      dtype=object)

In [4]:
# Verificar que las filas de ambos datasets coincidan fila por fila
DF_Uno.shape, DF_Dos.shape

((30641, 9), (30641, 15))

In [5]:
DF_Uno["Unnamed: 0"].duplicated().sum()

np.int64(29641)

In [6]:
DF_Dos["Unnamed: 0"].duplicated().sum()

np.int64(29641)

Ambos DataSets contienen la misma cantidad de filas, la diferencia son las columnas extras que contiene el DataSet 2 (DF_Dos). Se resalta que la columna "Unnamed: 0" contiene valores duplicados ya que el indice se renicia después de 999. Se uniran ambos datasets para evitar el problema de los valores duplicados se creará una columna en cada DatSet con un valor único y esa columna servirá para realizar el merge.

In [7]:
# Crear columna índice con valores únicos para DataSet Uno
DF_Uno["ID"] = range(1, len(DF_Uno)+1)

# Colocar la columna ID al inicio
columna = ["ID"] + [col for col in DF_Uno.columns if col != "ID"]
DF_Uno = DF_Uno[columna]

In [8]:
DF_Uno.head(10)

ID  Unnamed: 0  Gender EthnicGroup          ParentEduc     LunchType  \
0   1           0  female     group B   bachelor's degree      standard   
1   2           1  female     group C        some college      standard   
2   3           2  female     group B     master's degree      standard   
3   4           3    male     group A  associate's degree  free/reduced   
4   5           4    male     group C        some college      standard   
5   6           5  female     group B  associate's degree      standard   
6   7           6  female     group B        some college      standard   
7   8           7    male     group B        some college  free/reduced   
8   9           8    male     group D         high school  free/reduced   
9  10           9  female     group B         high school  free/reduced   

    TestPrep  MathScore  ReadingScore  WritingScore  
0       none         72            72            74  
1  completed         69            90            88  
2       none         90            95            93  
3       none         47            57            44  
4       none         76            78            75  
5       none         71            83            78  
6  completed         88            95            92  
7       none         40            43            39  
8  completed         64            64            67  
9       none         38            60            50

In [9]:
# Crear columna índice con valores únicos para DataSet Dos
DF_Dos["ID"] = range(1, len(DF_Dos)+1)

# Colocar la columna ID al inicio
columna = ["ID"] + [col for col in DF_Dos.columns if col != "ID"]
DF_Dos = DF_Dos[columna]

In [10]:
DF_Dos.head(10)

ID  Unnamed: 0  Gender EthnicGroup          ParentEduc     LunchType  \
0   1           0  female         NaN   bachelor's degree      standard   
1   2           1  female     group C        some college      standard   
2   3           2  female     group B     master's degree      standard   
3   4           3    male     group A  associate's degree  free/reduced   
4   5           4    male     group C        some college      standard   
5   6           5  female     group B  associate's degree      standard   
6   7           6  female     group B        some college      standard   
7   8           7    male     group B        some college  free/reduced   
8   9           8    male     group D         high school  free/reduced   
9  10           9  female     group B         high school  free/reduced   

    TestPrep ParentMaritalStatus PracticeSport IsFirstChild  NrSiblings  \
0       none             married     regularly          yes         3.0   
1        NaN             married     sometimes          yes         0.0   
2       none              single     sometimes          yes         4.0   
3       none             married         never           no         1.0   
4       none             married     sometimes          yes         0.0   
5       none             married     regularly          yes         1.0   
6  completed             widowed         never           no         1.0   
7       none             married     sometimes          yes         1.0   
8  completed              single     sometimes           no         3.0   
9       none             married     regularly          yes         NaN   

  TransportMeans WklyStudyHours  MathScore  ReadingScore  WritingScore  
0     school_bus            < 5         71            71            74  
1            NaN         5 - 10         69            90            88  
2     school_bus            < 5         87            93            91  
3            NaN         5 - 10         45            56            42  
4     school_bus         5 - 10         76            78            75  
5     school_bus         5 - 10         73            84            79  
6        private         5 - 10         85            93            89  
7        private           > 10         41            43            39  
8        private           > 10         65            64            68  
9        private            < 5         37            59            50

Las columnas a considerar del segundo DataFrame son:
- Medio De Transporte
- Estado Civil
- Practia De Deporte
- Horas De Estudio Semanalmente

In [50]:
NotasEstudiantes = pd.merge(
    DF_Uno,
    DF_Dos[["ID", "ParentMaritalStatus", "PracticeSport", "TransportMeans", "WklyStudyHours"]],
    on="ID",
    how="left"
)

NotasEstudiantes

ID  Unnamed: 0  Gender EthnicGroup          ParentEduc  \
0          1           0  female     group B   bachelor's degree   
1          2           1  female     group C        some college   
2          3           2  female     group B     master's degree   
3          4           3    male     group A  associate's degree   
4          5           4    male     group C        some college   
...      ...         ...     ...         ...                 ...   
30636  30637         995    male     group C    some high school   
30637  30638         996    male     group E  associate's degree   
30638  30639         997    male     group C        some college   
30639  30640         998    male     group A    some high school   
30640  30641         999  female     group D  associate's degree   

          LunchType   TestPrep  MathScore  ReadingScore  WritingScore  \
0          standard       none         72            72            74   
1          standard  completed         69            90            88   
2          standard       none         90            95            93   
3      free/reduced       none         47            57            44   
4          standard       none         76            78            75   
...             ...        ...        ...           ...           ...   
30636      standard       none         56            47            51   
30637  free/reduced       none         74            75            72   
30638      standard       none         36            29            27   
30639  free/reduced  completed         43            34            39   
30640      standard       none         52            68            66   

      ParentMaritalStatus PracticeSport TransportMeans WklyStudyHours  
0                 married     regularly     school_bus            < 5  
1                 married     sometimes            NaN         5 - 10  
2                  single     sometimes     school_bus            < 5  
3                 married         never            NaN         5 - 10  
4                 married     sometimes     school_bus         5 - 10  
...                   ...           ...            ...            ...  
30636              single     sometimes     school_bus         5 - 10  
30637              single     regularly        private         5 - 10  
30638             married     sometimes        private         5 - 10  
30639             married     regularly     school_bus         5 - 10  
30640             married         never     school_bus         5 - 10  

[30641 rows x 14 columns]

Se procederá a eliminar la columna "Unnamed: 0" y a cambiar los nombres de las columnas a español.

In [12]:
# Eliminar columna 'Unnamed: 0'
del NotasEstudiantes["Unnamed: 0"]

In [13]:
# Cambiar nombre de columnas a español
NotasEstudiantesDos = NotasEstudiantes.rename(columns=
                        {"Gender":"Género",
                         "EthnicGroup":"Grupo Étnico",
                         "ParentEduc":"Nivel Educativo Padres",
                         "LunchType":"Tipo De Almuerzo",
                         "TestPrep":"Preparación Para Examen",
                         "MathScore":"Nota Matemáticas",
                         "ReadingScore":"Nota Lectura",
                         "WritingScore":"Nota Escritura",
                         "ParentMaritalStatus":"Estado Civil Padres",
                         "PracticeSport":"Práctica Deportiva",
                         "TransportMeans":"Medio De Transporte",
                         "WklyStudyHours":"Horas De Estudio Semanal"})

NotasEstudiantesDos.head(3)

ID  Género Grupo Étnico Nivel Educativo Padres Tipo De Almuerzo  \
0   1  female      group B      bachelor's degree         standard   
1   2  female      group C           some college         standard   
2   3  female      group B        master's degree         standard   

  Preparación Para Examen  Nota Matemáticas  Nota Lectura  Nota Escritura  \
0                    none                72            72              74   
1               completed                69            90              88   
2                    none                90            95              93   

  Estado Civil Padres Práctica Deportiva Medio De Transporte  \
0             married          regularly          school_bus   
1             married          sometimes                 NaN   
2              single          sometimes          school_bus   

  Horas De Estudio Semanal  
0                      < 5  
1                   5 - 10  
2                      < 5

In [14]:
NotasEstudiantesDos.describe()

ID  Nota Matemáticas  Nota Lectura  Nota Escritura
count  30641.000000      30641.000000  30641.000000    30641.000000
mean   15321.000000         66.749355     69.624980       68.468327
std     8845.439136         15.206049     14.671572       15.307814
min        1.000000          0.000000     10.000000        5.000000
25%     7661.000000         56.000000     60.000000       58.000000
50%    15321.000000         67.000000     70.000000       69.000000
75%    22981.000000         78.000000     80.000000       79.000000
max    30641.000000        100.000000    100.000000      100.000000

- Todas las columnas númericas muestran 30,641 registros así que no hay valores NaN.
- Las notas promedio están entre 66 y 70, lo que indica que los alumnos regularmente aprueban, pero no con notas altas.
- La variabilidad es moderada ya que los estudiantes tienen notas variadas, no están concentradas en un solo nivel.
- Los valores mínimos encontrados son muy bajos, posiblemente haya datos atípicos.
- El 25% de los estudiantes tienen notas bajas o mediobajas.
- El 50% de los estudiantes tiene notas muy similares a la medía, por lo que la distribución es más o menos simétrica.
- Solo un 25% de los estudiantes se encuentra por arriba del promedio, lo que quiere decir que están en un rango alto.

**Estos valores y demás serán greaficados para una mejor visualización**

In [15]:
NotasEstudiantesDos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30641 entries, 0 to 30640
Data columns (total 13 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   ID                        30641 non-null  int64 
 1   Género                    30641 non-null  object
 2   Grupo Étnico              30641 non-null  object
 3   Nivel Educativo Padres    30641 non-null  object
 4   Tipo De Almuerzo          30641 non-null  object
 5   Preparación Para Examen   30641 non-null  object
 6   Nota Matemáticas          30641 non-null  int64 
 7   Nota Lectura              30641 non-null  int64 
 8   Nota Escritura            30641 non-null  int64 
 9   Estado Civil Padres       29451 non-null  object
 10  Práctica Deportiva        30010 non-null  object
 11  Medio De Transporte       27507 non-null  object
 12  Horas De Estudio Semanal  29686 non-null  object
dtypes: int64(4), object(9)
memory usage: 3.0+ MB


In [27]:
NotasEstudiantesDos[["Estado Civil Padres","Práctica Deportiva","Medio De Transporte", "Horas De Estudio Semanal"]].isnull().sum()

Estado Civil Padres         1190
Práctica Deportiva           631
Medio De Transporte         3134
Horas De Estudio Semanal     955
dtype: int64

Estas son las columnas que contienen valores nulos, esto se modificará para que no afecte en las visualizaciones:
 - Para la columna "Estado Civil Padres" los valores nulos se modificaran a not specified
 - Para la columna "Práctica Deportiva" los valores nulos se modificaran a sometimes debido a la alta cantidad de registros con estos datos
 - Para la columna "Medio De Transporte" los valores nulos se modificaran a walking 
 - Para la columna "Horas De Estudio Semanal" los valores nulos se modificaran a 5 - 10 debido a la alta cantidad de registros con estos datos

In [40]:
# Cambio de valores nulos
# 1. Columna Estado Civil Padres
NotasEstudiantesDos["Estado Civil Padres"] = (
    NotasEstudiantesDos["Estado Civil Padres"].fillna("not specified")
)

# 2. Columna Práctiva Deportiva
modaDeporte = NotasEstudiantesDos["Práctica Deportiva"].mode()[0]

NotasEstudiantesDos["Práctica Deportiva"] = (
    NotasEstudiantesDos["Práctica Deportiva"].fillna(modaDeporte)
)

# 3. Columna Medio De Transporte
NotasEstudiantesDos["Medio De Transporte"] = (
    NotasEstudiantesDos["Medio De Transporte"].fillna("walking")
)

# 4. Columna Horas De Estudio Semanal
modaEstudioSemanal = NotasEstudiantesDos["Horas De Estudio Semanal"].mode()[0]

NotasEstudiantesDos["Horas De Estudio Semanal"] = (
    NotasEstudiantesDos["Horas De Estudio Semanal"].fillna(modaEstudioSemanal)
)

In [41]:
NotasEstudiantesDos.isnull().sum()

ID                          0
Género                      0
Grupo Étnico                0
Nivel Educativo Padres      0
Tipo De Almuerzo            0
Preparación Para Examen     0
Nota Matemáticas            0
Nota Lectura                0
Nota Escritura              0
Estado Civil Padres         0
Práctica Deportiva          0
Medio De Transporte         0
Horas De Estudio Semanal    0
dtype: int64

El DataFrame ya no contiene valores nulos

# Datos Cálculados

In [17]:
Grupos = NotasEstudiantesDos.groupby("Grupo Étnico")["Grupo Étnico"].count()
print(f"Cantidad de alumnos pertenecientes a cada grupo\n{Grupos}")

Cantidad de alumnos pertenecientes a cada grupo
Grupo Étnico
group A    2359
group B    6192
group C    9816
group D    7986
group E    4288
Name: Grupo Étnico, dtype: int64


In [18]:
GruposMatematicaPromedio = NotasEstudiantesDos.groupby("Grupo Étnico")["Nota Matemáticas"].mean()
print(f"Promedio de notas por cada grupo para el examen de Matemáticas \n{round(GruposMatematicaPromedio, 1)}")

Promedio de notas por cada grupo para el examen de Matemáticas 
Grupo Étnico
group A    63.1
group B    63.7
group C    64.9
group D    67.8
group E    75.5
Name: Nota Matemáticas, dtype: float64


In [19]:
GruposLecturaPromedio = NotasEstudiantesDos.groupby("Grupo Étnico")["Nota Lectura"].mean()
print(f"Promedio de notas por cada grupo para el examen de Lectura \n{round(GruposLecturaPromedio, 1)}")

Promedio de notas por cada grupo para el examen de Lectura 
Grupo Étnico
group A    67.0
group B    67.6
group C    68.7
group D    70.6
group E    74.5
Name: Nota Lectura, dtype: float64


In [20]:
GruposEscrituraPromedio = NotasEstudiantesDos.groupby("Grupo Étnico")["Nota Escritura"].mean()
print(f"Promedio de notas por cada grupo para el examen de Escritura \n{round(GruposEscrituraPromedio, 1)}")

Promedio de notas por cada grupo para el examen de Escritura 
Grupo Étnico
group A    65.2
group B    66.0
group C    67.0
group D    70.8
group E    72.7
Name: Nota Escritura, dtype: float64


Se observa que el grupo A es el que tiene los promedios más bajos en los 3 examenes, esto se puede deber a varios factores:
- Tipo de transporte en el que se moviliza
- Horas de estudio a la semana invertidas

Se realizarán visualizaciones para llegar a una mejor conclusión.

# Guardar Dataframe Limpio

In [52]:
NotasEstudiantesDos.to_csv("../Data/NotasEstudiantesLimpio.csv", index=False)